pkg/middleware/auth_proxy/auth_proxy.go

```diff
@@ -0,0 +1,320 @@
+func (auth *AuthProxy) HasHeader() bool {
+	if len(auth.header) == 0 {
+		return false
+	}
+
+	return true
+}
+
+// IsAllowedIP compares presented IP with the whitelist one
+func (auth *AuthProxy) IsAllowedIP() (bool, *Error) {
+	ip := auth.ctx.Req.RemoteAddr
+
+	if len(strings.TrimSpace(auth.whitelistIP)) == 0 {
+		return true, nil
+	}
+
+	proxies := strings.Split(auth.whitelistIP, ",")
+	var proxyObjs []*net.IPNet
+	for _, proxy := range proxies {
+		result, err := coerceProxyAddress(proxy)
+		if err != nil {
+			return false, newError("Could not get the network", err)
+		}
+
+		proxyObjs = append(proxyObjs, result)
+	}
+
+	sourceIP, _, _ := net.SplitHostPort(ip)
+	sourceObj := net.ParseIP(sourceIP)
+
+	for _, proxyObj := range proxyObjs {
+		if proxyObj.Contains(sourceObj) {
+			return true, nil
+		}
+	}
+
+	err := fmt.Errorf(
+		"Request for user (%s) from %s is not from the authentication proxy", auth.header,
+		sourceIP,
+	)
+
+	return false, newError("Proxy authentication required", err)
+}
+
+// InCache checks if we have user in cache
+func (auth *AuthProxy) InCache() bool {
+	userID, _ := auth.GetUserIDViaCache()
+
+	if userID == 0 {
+		return false
+	}
+
+	return true
+}
+
+// getKey forms a key for the cache
+func (auth *AuthProxy) getKey() string {
+	return fmt.Sprintf(CachePrefix, auth.header)
+}
+
+// GetUserID gets user id with whatever means possible
+func (auth *AuthProxy) GetUserID() (int64, *Error) {
+	if auth.InCache() {
+
+		// Error here means absent cache - we don't need to handle that
+		id, _ := auth.GetUserIDViaCache()
+
+		return id, nil
+	}
+
+	if auth.ldapEnabled {
+		id, err := auth.GetUserIDViaLDAP()
+
+		if err == login.ErrInvalidCredentials {
+			return 0, newError("Proxy authentication required", login.ErrInvalidCredentials)
+		}
+
+		if err != nil {
+			return 0, newError("Failed to sync user", err)
+		}
+
+		return id, nil
+	}
+
+	id, err := auth.GetUserIDViaHeader()
+	if err != nil {
+		return 0, newError("Failed to login as user specified in auth proxy header", err)
+	}
+
+	return id, nil
+}
+
+// GetUserIDViaCache gets the user from cache
+func (auth *AuthProxy) GetUserIDViaCache() (int64, error) {
+	var (
+		cacheKey    = auth.getKey()
+		userID, err = auth.store.Get(cacheKey)
+	)
+
+	if err != nil {
+		return 0, err
+	}
+
+	return userID.(int64), nil
+}
+
+// GetUserIDViaLDAP gets user via LDAP request
+func (auth *AuthProxy) GetUserIDViaLDAP() (int64, *Error) {
+	query := &models.LoginUserQuery{
+		ReqContext: auth.ctx,
+		Username:   auth.header,
+	}
+
+	ldapCfg := login.LdapCfg
+	if len(ldapCfg.Servers) < 1 {
+		return 0, newError("No LDAP servers available", nil)
+	}
+
+	for _, server := range ldapCfg.Servers {
+		author := auth.LDAP(server)
+		if err := author.SyncUser(query); err != nil {
+			return 0, newError(err.Error(), nil)
+		}
+	}
+
+	return query.User.Id, nil
+}
+
+// GetUserIDViaHeader gets user from the header only
+func (auth *AuthProxy) GetUserIDViaHeader() (int64, error) {
```

Comment:
I guess it would be better to add more tests for this one as well as refactor this logic a bit too.
But wanna get feedback for the general approach first :)